include/aspect/particle/particle_handler.h

```diff
@@ -0,0 +1,258 @@
+ along with ASPECT; see the file LICENSE.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _aspect_particle_particle_handler_h
+#define _aspect_particle_particle_handler_h
+
+#include <aspect/global.h>
+#include <aspect/particle/particle.h>
+#include <aspect/particle/particle_accessor.h>
+#include <aspect/particle/particle_iterator.h>
+
+#include <aspect/particle/property_pool.h>
+
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/array_view.h>
+#include <deal.II/base/smartpointer.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    using namespace dealii;
+
+    template <int> class World;
+
+    /**
+     * This class manages the storage and handling of particles. It provides
+     * the data structures necessary to store particles efficiently, accessor
+     * functions to iterate over particles and find particles, and algorithms
+     * to distribute particles in parallel domains.
+     *
+     * @ingroup Particle
+     */
+    template <int dim, int spacedim=dim>
+    class ParticleHandler: public Subscriptor
+    {
+      public:
+        /**
+         * A type that can be used to iterate over all particles in the domain.
+         */
+        typedef ParticleIterator<dim,spacedim> particle_iterator;
+
+        /**
+         * Default constructor.
+         */
+        ParticleHandler();
+
+        /**
+         * Default constructor.
+         */
+        ParticleHandler(const parallel::distributed::Triangulation<dim,spacedim> &tria,
+                        MPI_Comm mpi_communicator);
+
+        /**
+         * Default destructor.
+         */
+        ~ParticleHandler();
+
+        /**
+         * Initialize the particle world.
+         */
+        void initialize();
+
+        /**
+         * Clear all particle related data.
+         */
+        void clear();
+
+        /**
+         * Return an iterator to the first particle.
+         */
+        ParticleHandler<dim,spacedim>::particle_iterator begin() const;
+
+        /**
+         * Return an iterator to the first particle.
+         */
+        ParticleHandler<dim,spacedim>::particle_iterator begin();
+
+        /**
+         * Return an iterator past the end of the particles.
+         */
+        ParticleHandler<dim,spacedim>::particle_iterator end() const;
+
+        /**
+         * Return an iterator past the end of the particles.
+         */
+        ParticleHandler<dim,spacedim>::particle_iterator end();
+
+        /**
+         * Return a pair of particle iterators that mark the begin and end of
+         * the particles in a particular cell. The last iterator is the first
+         * particle that is not longer in the cell.
+         */
+        std::pair<ParticleHandler<dim,spacedim>::particle_iterator,ParticleHandler<dim,spacedim>::particle_iterator>
+        particle_range_in_cell(const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell) const;
```

Comment:
I think it would be really cool if one could do something like
```
  for (auto particle : particle_handler.particle_range_in_cell (cell))
```
For this, the function needs to return a range. C++11 apparently doesn't have a range class, but boost has: http://www.boost.org/doc/libs/1_49_0/libs/range/doc/html/range/reference/utilities/iterator_range.html
Could we use this?